Added "ACL LOG" command new in Redis 6 RC2

2014BDuck opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `$ tox` pass with this change (including linting)?
- [x] Does travis tests pass with this change (enable it first in your forked repo and wait for the travis build to finish)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change
Hi. First of all thanks for checking this.
There is an "ACL LOG" command added in [Redis 6 RC2](https://github.com/antirez/redis/compare/6.0-rc1...6.0-rc2). I tried to add it to redis-py.

The "ACL LOG" command supports two kind of arguments: `ACL LOG [<count> | RESET]`. For `<count>`, it returns a acl log list. For `RESET`, it returns 'OK'. So I parsed it into:
```
The original response for 'ACL LOG':
[[b'count', 1, b'reason', b'command', b'context', b'toplevel', b'object', b'set', b'username', b'duck', b'age-seconds', b'1.958', b'client-info', b'id=3180 addr=127.0.0.1:59566 fd=9 name= age=105 idle=0 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=30 qbuf-free=32738 obl=0 oll=0 omem=0 events=r cmd=set user=duck']]

The redis-py output for 'ACL LOG':
[{'count': 1, 'reason': 'command', 'context': 'toplevel', 'object': 'set', 'username': 'duck', 'age-seconds': 1.958, 'client-info': {'id': 3180, 'addr': '127.0.0.1:59566', 'fd': '9', 'name': '', 'age': 105, 'idle': 0, 'flags': 'N', 'db': 0, 'sub': 0, 'psub': 0, 'multi': -1, 'qbuf': 30, 'qbuf-free': 32738, 'obl': 0, 'oll': 0, 'omem': 0, 'events': 'r', 'cmd': 'set', 'user': 'duck'}}]

-------------------- break --------------------

The original response for 'ACL LOG RESET':
b'OK'

The redis-py output for 'ACL LOG RESET':
True
```

Note that the wget link in `.travis.yml` is changed as well since the RC1 does not have a `LOG` sub-command for `ACL`.

Please take a look. It's my first PR for open-source project. Thank you.

andymccurdy commented:
I skimmed the code and this initially looks well done. Thanks!

Is there any documentation for `ACL LOG` on redis.io or elsewhere? I looked at the ACL section of the redis.io website and found no mention of this.

itamarhaber commented:
It has yet to be documented

2014BDuck commented:
> Is there any documentation for `ACL LOG` on redis.io or elsewhere?

Yes [the ACL doc](https://redis.io/topics/acl) is not updated.
This pull request is based on Antirez's [videos](https://www.youtube.com/watch?v=pSd6B_Jwv28) and source code and some redis-cli test myself.

2014BDuck commented:
The return type may change from list to boolean, based on what argument is used. I am thinking about splitting them into two methods like `acl_log()` and `acl_log_reset()`.
But that will not match the Redis command usage which is `ACL LOG [<count> | RESET]`.

2014BDuck commented:
Hi. Any idea about this one?
https://redis.io/commands/acl-log

RoeyPrat commented:
> The return type may change from list to boolean, based on what argument is used. I am thinking about splitting them into two methods like `acl_log()` and `acl_log_reset()`.
> But that will not match the Redis command usage which is `ACL LOG [<count> | RESET]`.

Hi @2014BDuck, great work!
I think it would be better to split the method into two methods as you've suggested.
`acl_log` and `acl_log_reset` have different use cases, as well as returning different types.

Also note you can drop the commit to change redis version in travis, as we are now using version 6.0.5 on master.

2014BDuck commented:
> > The return type may change from list to boolean, based on what argument is used. I am thinking about splitting them into two methods like `acl_log()` and `acl_log_reset()`.
> > But that will not match the Redis command usage which is `ACL LOG [<count> | RESET]`.
> 
> Hi @2014BDuck, great work!
> I think it would be better to split the method into two methods as you've suggested.
> `acl_log` and `acl_log_reset` have different use cases, as well as returning different types.
> 
> Also note you can drop the commit to change redis version in travis, as we are now using version 6.0.5 on master.

- For 6.0.5: Cool let me merge the master first so that old commit will be covered
- For seperated it into 2 method: OK Will do

andymccurdy commented:
@2014BDuck Awesome, thanks. Ping me when you get this split into two separate functions and we'll get it merged.

2014BDuck commented:
@andymccurdy Hi ptal one more time. Thanks

andymccurdy commented on tests/test_commands.py:
```diff
@@ -195,6 +195,39 @@ def teardown():
+                      keys=['cache:*'], nopass=True)
+        r.acl_log_reset()
+
+        r_test = redis.Redis(host='localhost', port=6379, db=9,
```
We don't want to hardcode connection info in a test. Instead, use the `_get_client` function from `conftest.py`. Something like:

```python
r_test = _get_client(redis.Redis, request, username=username)
```

2014BDuck replied:
Yep please check the comment below

2014BDuck commented:
Hi @andymccurdy 
I've added some the "assert" cases and changed few cases into "with" pattern, which is more accurate for the testing. Thanks for your guidence & learned a lot.

For the hardcoded connection:
```
r_test = redis.Redis(host='localhost', port=6379, db=9...
```
Yep it's not pretty. I tried to find out the correct pattern to new a different client before writting it but haven't found any example.

The get_client way `r_test = _get_client(redis.Redis, request, username=username)` : Seems like the deeper codes won't use the `username` variable in `**kwargs` but from the passed in `request` object:
```python
# This method passed redus_url string grabbed from request obj. And passed to "from_url" class method.
def _get_client(cls, request, single_connection_client=True, **kwargs):
    redis_url = request.config.getoption("--redis-url")
    client = cls.from_url(redis_url, **kwargs)
    .....


# This method will parse url and use the username in url.
    def from_url(cls, url, db=None, decode_components=False, **kwargs):
        url = urlparse(url)
        url_options = {}

        ...
        # Here. The username is token directly from the url we passed in.
        if decode_components:
            username = unquote(url.username) if url.username else None
            ...
        else:
            username = url.username or None
            ...
```

So it's not working using `r_test = _get_client(redis.Redis, request, username=username)`. I am thinking about: 
- Altering the attribute of request obj ? like `request.config.option.redis_url = ''redis://xxx:localhost:6379/9'`. But I think it looks more weird than building a new client with `redis.Redis(host='localhost', port=6379, db=9)`
- Altering the deeper method `from_url` using username in `**kwargs` if existed. Otherwise following the current pattern.

Actually both of them sound worse from my point of view. Would like to hear more advice from your side. Many thanks.

2014BDuck commented on tests/test_commands.py:
```diff
@@ -195,6 +195,39 @@ def teardown():
+        r_test.set("cache:0", 1)
+        r_test.get("cache:0")
+        # Invalid key
+        try:
```
Done. Thanks.

2014BDuck commented on tests/test_commands.py:
```diff
@@ -195,6 +195,39 @@ def teardown():
+        except exceptions.NoPermissionError:
+            pass
+        # Invalid operation
+        try:
```
Done. Thanks.

2014BDuck commented on tests/test_commands.py:
```diff
@@ -195,6 +195,39 @@ def teardown():
+            pass
+
+        assert len(r.acl_log()) == 2
+        assert len(r.acl_log(count=1)) == 1
```
Added few return type assertion now.

2014BDuck commented:
@andymccurdy Hi. May I request for a review and your suggestion again?

Two things after I merged the current master branch.
- The coverage test is not passed due to...? Looks like it's not caused by my changes? Should I re-open a new PR or any suggestion?
- The hardcoded connection. If we need to pass the username as connection param, it would require additional changes on couple methods, which is not graceful enough from my point of view.

andymccurdy commented:
@2014BDuck I fixed this in the acl-log branch here: https://github.com/andymccurdy/redis-py/tree/acl-log

I believe that branch is good to go. Let me know if you see anything off. If you want to merge it back into this PR I'll then merge your PR to master. Alternatively I can just merge the acl-log branch to master. Let me know what you prefer.

2014BDuck commented:
> Alternatively I can just merge the acl-log branch to master

Yes please. 

And sorry for wasting your time and effort on guiding. Very appreciated.

andymccurdy commented:
Merged. Thanks!
